go-controller/pkg/libovsdb/ops/sample.go

```diff
@@ -0,0 +1,170 @@
+	return collectors, err
+}
+
+func CreateOrUpdateSamplingAppsOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, samplingApps ...*nbdb.SamplingApp) ([]libovsdb.Operation, error) {
+	opModels := make([]operationModel, 0, len(samplingApps))
+	for i := range samplingApps {
+		// can't use i in the predicate, for loop replaces it in-memory
+		samplingApp := samplingApps[i]
+		opModel := operationModel{
+			Model:          samplingApp,
+			OnModelUpdates: onModelUpdatesAllNonDefault(),
+			ErrNotFound:    false,
+			BulkOp:         false,
+		}
+		opModels = append(opModels, opModel)
+	}
+
+	modelClient := newModelClient(nbClient)
+	return modelClient.CreateOrUpdateOps(ops, opModels...)
+}
+
+func DeleteSamplingAppsWithPredicate(nbClient libovsdbclient.Client, p func(collector *nbdb.SamplingApp) bool) error {
+	opModel := operationModel{
+		Model:          &nbdb.SamplingApp{},
+		ModelPredicate: p,
+		ErrNotFound:    false,
+		BulkOp:         true,
+	}
+	m := newModelClient(nbClient)
+	return m.Delete(opModel)
+}
+
+func FindSample(nbClient libovsdbclient.Client, sampleMetadata int) (*nbdb.Sample, error) {
+	sample := &nbdb.Sample{
+		Metadata: sampleMetadata,
+	}
+	opModel := operationModel{
+		Model:       sample,
+		ErrNotFound: true,
+		BulkOp:      false,
+	}
+	modelClient := newModelClient(nbClient)
+	err := modelClient.Lookup(opModel)
+	if err != nil {
+		return nil, err
+	}
+	return sample, err
+}
+
+type SampleFeature int
+
+const (
+	EgressFirewall SampleFeature = iota
+	NetworkPolicy
+	AdminNetworkPolicy
+	Multicast
+	UDNIsolation
+)
+
+// SamplingConfig is used to configure sampling for different db objects.
+type SamplingConfig struct {
+	featureCollectors map[SampleFeature][]string
+}
+
+func NewSamplingConfig(featureCollectors map[SampleFeature][]string) *SamplingConfig {
+	return &SamplingConfig{
+		featureCollectors: featureCollectors,
+	}
+}
+
+func addSample(c *SamplingConfig, opModels []operationModel, model model.Model) []operationModel {
+	if c == nil {
+		return opModels
+	}
+	switch t := model.(type) {
+	case *nbdb.ACL:
+		return createOrUpdateSampleForACL(opModels, c, t)
+	}
+	return opModels
+}
+
+// createOrUpdateSampleForACL should be called before acl operationModel is appended to opModels.
+func createOrUpdateSampleForACL(opModels []operationModel, c *SamplingConfig, acl *nbdb.ACL) []operationModel {
+	collectors := c.featureCollectors[getACLSampleFeature(acl)]
+	if len(collectors) == 0 {
+		return opModels
+	}
+	sample := &nbdb.Sample{
+		Collectors: collectors,
+		// 32 bits
+		Metadata: int(GetSampleID(acl)),
+	}
+	opModel := operationModel{
+		Model: sample,
+		DoAfter: func() {
+			acl.SampleEst = &sample.UUID
+			acl.SampleNew = &sample.UUID
+		},
+		OnModelUpdates: []interface{}{&sample.Collectors},
+		ErrNotFound:    false,
+		BulkOp:         false,
+	}
+	opModels = append(opModels, opModel)
+	return opModels
+}
+
+func GetSampleID(dbObj hasExternalIDs) uint32 {
```

Comment:
Maybe I'm misunderstanding how this works but shouldn't the sample ID be unique per ACL?  Doesn't this function have a (small) chance of collision?

Reply:
yes, there is always a small chance of collision, but I don't see any other option that would guarantee consistent hashing and no collision.
this function is inspired by https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/pkg/util/util.go#L253 that we used before to generate address set and port group names.

Reply:
OK, I guess we have to live with that for now.